compiler/rustc_data_structures/src/tagged_ptr/copy.rs

```diff
@@ -1,78 +1,82 @@
 use super::{Pointer, Tag};
 use crate::stable_hasher::{HashStable, StableHasher};
 use std::fmt;
+use std::hash::{Hash, Hasher};
 use std::marker::PhantomData;
+use std::mem::ManuallyDrop;
 use std::num::NonZeroUsize;
+use std::ops::{Deref, DerefMut};
+use std::ptr::NonNull;
 
-/// A `Copy` TaggedPtr.
+/// A [`Copy`] tagged pointer.
 ///
-/// You should use this instead of the `TaggedPtr` type in all cases where
-/// `P: Copy`.
+/// This is essentially `{ pointer: P, tag: T }` packed in a single pointer.
+///
+/// You should use this instead of the [`TaggedPtr`] type in all cases where
+/// `P` implements [`Copy`].
 ///
 /// If `COMPARE_PACKED` is true, then the pointers will be compared and hashed without
-/// unpacking. Otherwise we don't implement PartialEq/Eq/Hash; if you want that,
-/// wrap the TaggedPtr.
+/// unpacking. Otherwise we don't implement [`PartialEq`], [`Eq`] and [`Hash`];
+/// if you want that, wrap the [`CopyTaggedPtr`].
+///
+/// [`TaggedPtr`]: crate::tagged_ptr::TaggedPtr
 pub struct CopyTaggedPtr<P, T, const COMPARE_PACKED: bool>
 where
     P: Pointer,
     T: Tag,
 {
-    packed: NonZeroUsize,
-    data: PhantomData<(P, T)>,
-}
-
-impl<P, T, const COMPARE_PACKED: bool> Copy for CopyTaggedPtr<P, T, COMPARE_PACKED>
-where
-    P: Pointer,
-    T: Tag,
-    P: Copy,
-{
-}
-
-impl<P, T, const COMPARE_PACKED: bool> Clone for CopyTaggedPtr<P, T, COMPARE_PACKED>
-where
-    P: Pointer,
-    T: Tag,
-    P: Copy,
-{
-    fn clone(&self) -> Self {
-        *self
-    }
+    /// This is semantically a pair of `pointer: P` and `tag: T` fields,
+    /// however we pack them in a single pointer, to save space.
+    ///
+    /// We pack the tag into the **most**-significant bits of the pointer to
+    /// ease retrieval of the value. A left shift is a multiplication and
+    /// those are embeddable in instruction encoding, for example:
+    ///
+    /// ```asm
+    /// // (https://godbolt.org/z/jqcYPWEr3)
+    /// example::shift_read3:
+    ///     mov     eax, dword ptr [8*rdi]
+    ///     ret
+    ///
+    /// example::mask_read3:
```

Comment:
You don't need to elaborate as much for it, but a quick check that this isn't totally destroying aarch64 perf would be useful (not that im worried about it being true)

Reply:
On aarch64 both approaches seem to [produce](https://godbolt.org/z/j1a434dqn) basically the same code, the only difference being the operation (shift vs and):
```asm
example::shift_read3:
        lsl     x8, x0, #3
        ldr     w0, [x8]
        ret

example::mask_read3:
        and     x8, x0, #0xfffffffffffffff8
        ldr     w0, [x8]
        ret
```
It's not really clear if `lsl` and `and` have different perf implication (this [llvm-mca comparison](https://godbolt.org/z/j14GaMar9) suggests that `and` is slightly better, because it doesn't pressure `CyUnitIS`, but it's unclear how to quantify that; the "Total Cycles" and "Total uOps" are the same, so I'd suggest this difference does not matter (probably (also I've spend too much time trying to make llvm-mca work on aarch64 and just don't want to think about it anymore))).

Reply:
that looks good enough, I don't think thinking too much about the cycles here matters in the end, thanks.